security/advancedtls/advancedtls.go

```diff
@@ -550,9 +556,15 @@ func (c *advancedTLSCreds) OverrideServerName(serverNameOverride string) error {
 //  1. does not have a good support on root cert reloading.
 //  2. will ignore basic certificate check when setting InsecureSkipVerify
 //     to true.
+//
+// peerVerifiedChains(output param): verified chain of certs from leaf to the
+// trust cert that the peer trusts.
+//  1. For server it is, client certs + Root ca that the server trusts
+//  2. For client it is, server certs + Root ca that the client trusts
 func buildVerifyFunc(c *advancedTLSCreds,
 	serverName string,
-	rawConn net.Conn) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	rawConn net.Conn,
+	peerVerifiedChains *[][]*x509.Certificate) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
```

Comment:
A pointer to a slice of slices of pointers to x509 certs?

Would it make more sense to return a second value that is `[][]*x509.Certificate` instead?

Also should we define a type / type alias for `[][]*x509.Certificate`?

Reply:
to be honest, I couldn't unpack all the syntactic complexity around that to make it work. Any help would be appreciated.

Reply:
@dfawley 

   cfg := credinternal.CloneTLSConfig(c.config)
   peerVerifiedChains := [][]*x509.Certificate{}
   cfg.VerifyPeerCertificate = buildVerifyFunc(c, "", rawConn, &peerVerifiedChains)
   conn := tls.Server(rawConn, cfg)
   if err := conn.Handshake(); err != nil {
      conn.Close()
      return nil, nil, err
   }

buildVerifyFunc() is returning a function that gets stored in cfg.VerifyPeerCertificate and later gets called as part of conn.Handshake(). Given this structure returning a second parameter would mean that we have to modify the prevalent use of the type "func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error" and that quickly becomes a big change in itself. 

Reg "type / type alias for [][]*x509.Certificate?" I can define it and replace in advancedtls.go where possible, does this name sound reasonable? (sorry, trying to minimize back and forth and want to close this soon)

type CertChains [][]*x509.Certificate

Reply:
@gtcooke94 - thoughts?

Reply:
We don't really have too much flexibility in adding return values - we build this verification func then set crypto/tls's `VerifyPeerCertificate` to it https://pkg.go.dev/crypto/tls

Reply:
As far as I can tell this doesn't involve anything async? Instead of passing the variable directly we could do something with channels? But given the constraints of the callback structure, I think something like this is how we'll have to get it out unfortunately.